src/policy/space.rs

```diff
@@ -229,7 +219,16 @@ pub trait Space<VM: VMBinding>: 'static + SFT + Sync + Downcast {
 
     fn address_in_space(&self, start: Address) -> bool {
         if !self.common().descriptor.is_contiguous() {
-            self.common().vm_map().get_descriptor_for_address(start) == self.common().descriptor
+            // If the space is not contiguous, we use `SFT_MAP` to check.
+            //
+            // Note: When comparing references (`&T`) using `==`, it compares by values.
+            // To compare by identities, we need to compare pointers.
+            //
+            // TODO: For `DenseChunkMap`, it is possible to reduce one level of dereferencing by
+            // letting each space remember its index in `DenseChunkMap::index_map`.
+            let self_ptr = self as *const _ as *const ();
+            let other_ptr = SFT_MAP.get_checked(start) as *const _ as *const ();
```

Comment:
Better to use `as *const Self` instead of `*const _`. We had a bug https://github.com/mmtk/mmtk-core/pull/750 where if the return type of `get_checked` is changed, the `*const _` cast may still compile but can sliently fail the pointer comparison later.

Same to L229 as well.

Reply:
Yes.  We can do that on line 229.  But on L230,  the returned SFT instance may not have the same type as Self.

But I think a deeper problem is that `SFT_MAP.get_checked(start) as *const _` is a `*const dyn SFT`, which is 16 bytes long.  But `*const ()` is 8 bytes long. Embarrassingly, neither the Rust reference nor the Rustonomicon specify the semantics of casting pointer to pointer.  This means casting pointers this way is really no better than loading only half of the 128-bit `&dyn` from the SFT, as what build4 did.

Reply:
> Embarrassingly, neither the Rust reference nor the Rustonomicon specify the semantics of casting pointer to pointer.

Not sure I follow. Doesn't the first rule [here](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/casting-between-types.html#pointer-casts) describe when it's allowed? Or do you mean the semantics of then deferencing the pointer, which is most likely undefined

Reply:
> Not sure I follow. Doesn't the first rule [here](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/casting-between-types.html#pointer-casts) describe when it's allowed? Or do you mean the semantics of then deferencing the pointer, which is most likely undefined

The link you provided points to an ancient version of the Rust Book for v1.25.  The newest version of the Rust Book does not contain that section.

I did mean the casting itself.  It is not an no-op bits-preserving cast like `transmute`, as it changes the number of bits.

Reply:
Ah right. That was the first link on Google for me when I searched for "raw pointer casting Rust".